lib/hotel.rb

```diff
@@ -0,0 +1,139 @@
+      available_rooms = check_avail_rooms_for(check_in, check_out)
+      if rooms_available?(available_rooms) # Will raise an ArgumentError for UI
+        reserved_room_nums = available_rooms[0..num_of_rooms - 1] # Takes the first 5 rooms from all available rooms
+        new_block = Block.new(reserved_for, check_in, check_out, reserved_room_nums)
+        # Return only one block reservation and pass it into the @all_reservations array so that they cannot access the rooms
+        @all_reservations << new_block
+        @block_reservations << new_block
+        return new_block # Returns all reserved rooms numbers for a block
+      end
+    end
+
+    def find_block(reserved_name)
+      valid?(reserved_name)
+      @block_reservations.each do |block|
+        if block.reserved_for == reserved_name
+          return block
+        end
+      end
+      raise ArgumentError.new("No block was found")
+    end # def
+
+    # As an administrator, I can reserve a room from within a block of rooms
+    def reserve_room_in_block(reserved_name, num_to_book)
+      valid?(reserved_name)
+      found_block = find_block(reserved_name)
+      check_num_input(found_block, num_to_book) # Raise ArgumentError to UI
+      avail_rooms = avail_rooms_in_block(reserved_name) # Check which room numbers are available in the block
+      now_reserved_in_block = avail_rooms[0..num_to_book - 1] # Book number of requested rooms in the current block
+      remaining_rooms = avail_rooms - now_reserved_in_block # Remaining rooms in requested block
+      found_block.update_rooms_in_block(remaining_rooms, now_reserved_in_block) # Updates available and reserved rooms in the requested block
+    end
+
+    # As an administrator, I can check whether a given block has any rooms available
+    def avail_rooms_in_block(reserved_name)
+      valid?(reserved_name)
+      found_block = find_block(reserved_name)
+      if found_block.avail_block_rooms.empty?
+        #TODO: CHANGE THIS INTO A DIFFERENT ERROR
+        raise ArgumentError.new("There are no more rooms available to reserve for this block")
+      else
+        return found_block.avail_block_rooms
+      end
+    end # def
+
+    # I can view a list of rooms that are not reserved for a given date range
+    def check_avail_rooms_for(next_check_in, next_check_out) # WAVE 2
+      next_reservation = DateRange.new(next_check_in, next_check_out)
+      valid_dates?(next_reservation)
+      existing_booked_rooms = next_reservation.overlap?(@all_reservations)
+      existing_booked_rooms.empty? ? @rooms : @rooms - existing_booked_rooms
+    end
+
+    # I can access the list of all_reservations for a specific date
+    def all_reservations_on(date)
+      current_reservations = []
+      date = Date.parse(date)
+      @all_reservations.each do |reservation|
+        current_reservations << reservation if date >= reservation.check_in && date < reservation.check_out
```

Comment:
Here you do a bunch of math on the members of a `DateRange` object. This is usually a good indicator that a method is warranted. Even if you only call it once, it will make your code easier to maintain, since you know exactly where this calculation is made, and to read:

```ruby
current_reservations << reservation if reservation.dates.include?(date)
# instead of
current_reservations << reservation if date >= reservation.check_in && date < reservation.check_out
```